File: StringToINTERCAL.py
#!/usr/bin/env python
"""
StringToINTERCAL.py

Converts a string to an INTERCAL script that will output said string.
usage: python StringToINTERCAL.py your string here

https://www.maxodev.org
https://github.com/Maxoplata/StringToINTERCAL
"""

import sys

__author__ = "Maxamilian Demian"
__email__ = "max@maxdemian.com"

# class definition
class StringToINTERCAL:
	politeCount = 0

	def politeLine(self, line):
		if self.politeCount == 3:
			self.politeCount = 0

			return 'PLEASE {}\n'.format(line)

		self.politeCount += 1

		return 'DO {}\n'.format(line)

	def leadingZeros(self, dec):
		count = len(dec)

		if count < 8:
			dec = ('0' * (8 - count)) + dec

		return dec

	def convertToINTERCAL(self, string):
		# reset politeCount
		self.politeCount = 0

		ret = self.politeLine(',1 <- #{}'.format(len(string)))

		lastCharLoc = 256
		for i in range(0, len(string)):
			charLoc = int(self.leadingZeros('{0:b}'.format(ord(string[i])))[::-1], 2)

			movePosition = 0

			if charLoc < lastCharLoc:
				movePosition = (lastCharLoc - charLoc)
			elif charLoc > lastCharLoc:
				movePosition = (256 - charLoc) + lastCharLoc

			lastCharLoc -= movePosition
			if lastCharLoc < 1:
				lastCharLoc = 256 + lastCharLoc

			ret += self.politeLine(',1 SUB #{} <- #{}'.format((i + 1), movePosition))

		ret += self.politeLine('READ OUT ,1')
		ret += self.politeLine('GIVE UP')

		return ret

# 1337 codez
if len(sys.argv) > 1:
	inputString = ' '.join(sys.argv[1:])

	myINTERCAL = StringToINTERCAL()

	print(myINTERCAL.convertToINTERCAL(inputString))

File: StringToINTERCAL.rb
#!/usr/bin/ruby
# StringToINTERCAL.rb
#
# Converts a string to an INTERCAL script that will output said string.
# usage: ruby StringToINTERCAL.rb your string here
#
# @author Maxamilian Demian <max@maxdemian.com>
# @link https://www.maxodev.org
# @link https://github.com/Maxoplata/StringToINTERCAL

# class definition
class StringToINTERCAL
	def initialize()
		@politeCount = 0
	end

	def politeLine(line)
		if @politeCount === 3
			@politeCount = 0

			return "PLEASE #{line}\n"
		end

		@politeCount += 1

		return "DO #{line}\n"
	end

	def leadingZeros(dec)
		count = dec.length

		if count < 8
			dec = ('0' * (8 - count)) + dec
		end

		return dec
	end

	def convertToINTERCAL(string)
		# reset politeCount
		@politeCount = 0

		chars = string.split('')

		ret = self.politeLine(',1 <- #' + chars.count.to_s)

		lastCharLoc = 256
		chars.each_with_index {|char, i|
			charLoc = self.leadingZeros(char.ord.to_s(2)).reverse.to_i(2)

			movePosition = 0

			if charLoc < lastCharLoc
				movePosition = (lastCharLoc - charLoc)
			elsif charLoc > lastCharLoc
				movePosition = (256 - charLoc) + lastCharLoc
			end

			lastCharLoc -= movePosition
			if lastCharLoc < 1
				lastCharLoc = 256 + lastCharLoc
			end

			ret += self.politeLine(',1 SUB #' + (i + 1).to_s + " <- ##{movePosition}")
		}

		ret += self.politeLine('READ OUT ,1')
		ret += self.politeLine('GIVE UP')

		return ret
	end
end

# 1337 codez
if ARGV.count > 0
	inputString = ARGV.join(' ')

	puts StringToINTERCAL.new.convertToINTERCAL(inputString)
end
